core/engine/engine.go

```diff
@@ -1,99 +1,166 @@
 )
 
 var log = logging.Logger("reference-provider")
 
 var _ core.Interface = &Engine{}
 
+const (
+	// NOTE: For now this is an IPLD-aware reference provider, wo
+	// we need to force GraphSupport. This may change in the future
+	// and we may pass this as an option.
+	graphSupport = true
+)
+
 // Engine is an implementation of the core reference provider interface.
 type Engine struct {
+	// Provider's privateKey
+	privKey crypto.PrivKey
 	// Host running the provider process.
 	host host.Host
+	// Linksystem used for reference provider.
+	lsys ipld.LinkSystem
 	// Datastore used for persistence of different assets
 	// (advertisements, indexed data, etc.).
 	ds datastore.Batching
-	// Local index with all provided data.
-	// NOTE: We may want to wrap indexer.Interface into datastore interface
-	// to be able to use it as a datastore right away.
-	index indexer.Interface
-	// go-legs subscriber for IPLD-aware ingestion
-	ipldIngest legs.LegPublisher
+	lp legs.LegPublisher
 	// pubsubtopic where the provider will push advertisements
 	pubSubTopic string
+
+	// Local indexer with all provided data.
+	// NOTE: It may be the case that providers want to keep a local index with all
+	// provider data, and they want to store it with go-indexer-core. In this case
+	// we may need to wrap go-indexer-core so linkSystems can use it.
+	//indexer indexer.Interface
 }
 
 // New creates a reference provider engine with the corresponding config.
-func New(ctx context.Context, host host.Host, ds datastore.Batching, index indexer.Interface, pubSubTopic string) (*Engine, error) {
+func New(ctx context.Context,
+	cfg config.Config, host host.Host, ds datastore.Batching, index indexer.Interface,
+	pubSubTopic string) (*Engine, error) {
+
 	log.Debugw("Starting new reference provider engine")
 	lsys := mkLinkSystem(ds)
 	lp, err := legs.NewPublisher(ctx, ds, host, pubSubTopic, lsys)
 	if err != nil {
 		return nil, err
 	}
+	// TODO(security): We shouldn't keep the privkey decoded here.
+	// We should probably unlock it and lock it every time we need it.
+	// Once we start encrypting the key locally.
+	privKey, err := cfg.Identity.DecodePrivateKey("")
+	if err != nil {
+		return nil, err
+	}
 	return &Engine{
 		host:        host,
 		ds:          ds,
-		index:       index,
-		ipldIngest:  lp,
+		lp:          lp,
+		privKey:     privKey,
 		pubSubTopic: pubSubTopic,
 	}, nil
 }
 
-// TODO: This will probably change once we start implementing the actual ingestion protocol.
-func mkLinkSystem(ds datastore.Batching) ipld.LinkSystem {
-	lsys := cidlink.DefaultLinkSystem()
-	lsys.StorageReadOpener = func(lctx ipld.LinkContext, lnk ipld.Link) (io.Reader, error) {
-		c := lnk.(cidlink.Link).Cid
-		val, err := ds.Get(datastore.NewKey(c.String()))
-		if err != nil {
-			return nil, err
-		}
-		return bytes.NewBuffer(val), nil
+func (e *Engine) PublishLocal(ctx context.Context, adv core.Advertisement) (cid.Cid, error) {
+	return e.publishLocal(ctx, adv)
```

Comment:
I am curious what's the rationale for this refactor? func signatures are identical.